### PR TITLE
added few mixin features

### PR DESCRIPTION
**Pull request by Najidnadri**
CHANGE LOG
- added `mixin:class`
- remove `block` style in mixin:show if signal is true
- added `mixin:nshow` 
- added `DomTokenList` in hirola-core cargo.toml
- added `classes~` macro

Feel free to give any comments :)

### CODE REVIEW

File: crates/hirola-core/src/mixins.rs [hunk 1]
```diff
@@ -163,3 +213,256 @@ pub mod model {
...
+/// 
+///     let callback = Box::new(cloned!(img_src, img_alt) => move |_e: Event| {
+///         img_src.set("./dog.jpg".to_string());
+///         img_alt.set("dog image here".to_string());
+///     });
+///     
+///     html! {
+///         <div>
+///             <img mixin:src=&src(&img_src) mixin:alt=&alt(&img_alt) />
+///             <button on:click=callback)>"Change to dog"</button>
```
Comment (geofmureithi) [inline crates/hirola-core/src/mixins.rs]: This wont work

File: crates/hirola-core/src/mixins.rs [hunk 2]
```diff
@@ -163,3 +213,256 @@ pub mod model {
...
+///         </div>
+///     }
+/// }
+/// ```
+*/
+pub fn alt<'a>(signal: &'a Signal<String>) -> Box<dyn Fn(DomNode) -> () + 'a> {
```
Comment (geofmureithi) [inline crates/hirola-core/src/mixins.rs]: Why do we need this?
What is the difference with just using 
```
<img alt={&signal} />
```

File: crates/hirola-core/src/mixins.rs [hunk 3]
```diff
@@ -119,6 +145,30 @@ pub fn show(shown: &Signal<bool>) -> Box<dyn Fn(DomNode)> {
     Box::new(cb)
 }
 
+/// Opposite to `mixin:show`, element will be display if the signal is false
+pub fn nshow(shown: &Signal<bool>) -> Box<dyn Fn(DomNode)> {
```
Comment (geofmureithi) [inline crates/hirola-core/src/mixins.rs:151]: This makes good sense 👏

File: crates/hirola-core/src/mixins.rs [hunk 4]
```diff
@@ -163,3 +213,256 @@ pub mod model {
...
+///             .yellow-box {
+///                 background-color: yellow;
+///             }
+///             "##</style>
+///             <button on:click=change_class>"Click to change colour"</button>
+///         </div>
+///     }
+/// }
+/// ```
+pub fn class<'a>(signal: &'a Signal<Vec<String>>) -> Box<dyn Fn(DomNode) -> () + 'a> {
```
Comment (geofmureithi) [inline crates/hirola-core/src/mixins.rs:270]: This also makes good sense, but am curious if a macro would make more sense.

File: crates/hirola-core/src/mixins.rs [hunk 5]
```diff
@@ -163,3 +213,256 @@ pub mod model {
...
+///     
+///     html! {
+///         <div>
+///             <img mixin:src=&src(&state) />
+///             <button on:click=move |_| state.set("./dog.jpg".to_string())>"Change to dog"</button>
+///         </div>
+///     }
+/// }
+/// ```
+pub fn src<'a>(signal: &'a Signal<String>) -> Box<dyn Fn(DomNode) -> () + 'a> {
```
Comment (geofmureithi) [inline crates/hirola-core/src/mixins.rs]: This doesn't make much sense. I havent tried it but I think it should work without a mixin ie:
```rust 
/// pub fn reactive_img(_app: &HirolaApp) -> Dom {
///     let state = Signal::new(String::from("./cat.jpg"));
///     
///     html! {
///         <div>
///             <img src=&state.get() />
///             <button on:click=move |_| state.set("./dog.jpg".to_string())>"Change to dog"</button>
///         </div>
///     }
/// }

File: crates/hirola-core/src/mixins.rs [hunk 6]
```diff
@@ -163,3 +213,256 @@ pub mod model {
...
+///     
+///     html! {
+///         <div>
+///             <a mixin:href=&href(&state) >"Send me to somewhere not here"</a>
+///             <button on:click=move |_| state.set("https://youtube.com".to_string())>"Change to link to youtube"</button>
+///         </div>
+///     }
+/// }
+/// ```
+pub fn href<'a>(signal: &'a Signal<String>) -> Box<dyn Fn(DomNode) -> () + 'a> {
```
Comment (geofmureithi) [inline crates/hirola-core/src/mixins.rs]: I dont think we need a mixin for href too just like alt and src

**Review by geofmureithi — CHANGES_REQUESTED** [review]
I am unsure about:

- src
- alt
- href
- disabled
- id

I haven't tested but will do so tomorrow. Feel free to get back on your opinions.

File: crates/hirola-core/src/mixins.rs (hunk 2 above)
Comment (Najidnadri) [reply crates/hirola-core/src/mixins.rs]: You are definitely right. Never thought of doing that way before. Hence I believe, we didnt really need:
- mixin:src
- mixin:href
- mixin:alt
- mixin:id

File: crates/hirola-core/src/mixins.rs (hunk 4 above)
Comment (Najidnadri) [reply crates/hirola-core/src/mixins.rs:270]: Interesting, can I know where do you think we can call this macro from?

**Comment by Najidnadri** [comment]
Need your opinion on this two.

1. 
```
<div>
      <button disabled="true">button1</button>
      <button disabled="false">button2</button>
</div>
```
Both button will be disabled even if the value is `false` as long as the attribute `disabled` exist. One way we can handle this is by creating a mixin to remove or add the attribute. 

Hence, do you think we need mixin for this, or do you have any workaround for this issue?


2.
```
pub fn page(_app: &HirolaApp) -> Dom {
    let button1 = Signal::new(false);
    let button2 = Signal::new(false);
    
    //method 1 *will not work* 
    let shown = create_memo(cloned!((button1, button2) => move || {
        *button1.get() && *button2.get()
    }));
    
    //method2 *will work*
    let shown = Signal::new(false);
    create_effect(cloned!((button1, button2, shown) => move || {
        shown.set(*button1.get() && *button2.get());
    }));

    html!{
        <div>
            <div>
                <h1 mixin:show=&show(&shown)>"Show Me if both button is true!"</h1>
                <button on:click=move |_| button1.set(!*button1.get())} >"Button 1"</button>
                <button on:click=move |_| button2.set(!*button2.get()) >"Button 2"</button>
            </div>
        </div>
    }
}
```
`method 1` will not work because `create_memo` returns a `StateHandle`. However `mixin:show` need a `Signal` as the argument. 
Some of possible solution I can think of
- `create_memo` returns a`Signal` instead of `StateHandle`.
- create a `from_statehandle()` method for `Signal`
- developer should go with `Method 2`

(Of course do clarify if you have any other workaround on this one)

File: crates/hirola-core/src/mixins.rs [hunk 7]
```diff
@@ -163,3 +213,256 @@ pub mod model {
...
     }
 }
+
+
+///Mixin for reactive classes
+/// 
+/// ## Example
+/// ```
+/// pub fn yellow_or_black_box(_app: &HirolaApp) -> Dom {
+///     let state = Signal::new(vec!["black-box".to_string()]);
```
Comment (geofmureithi) [inline crates/hirola-core/src/mixins.rs:244]: Something similar to 
```rust
use semester::{classes, Classes as _};

let classes = classes!(
    "always",
    "yup": 10 == 10,
    "nope": 10 == 15,
);

assert_eq!(classes.render(), "always yup");
```
See https://lib.rs/crates/semester

Then one could just pass `Signal` in the bool parts right?

File: crates/hirola-core/src/mixins.rs (hunk 4 above)
Comment (geofmureithi) [reply crates/hirola-core/src/mixins.rs:270]: https://github.com/geofmureithi/hirola/pull/6/files/e165ed0641d259ff12b4e86f58095072453668d4#r998411264

File: crates/hirola-core/src/mixins.rs (hunk 7 above)
Comment (Najidnadri) [reply crates/hirola-core/src/mixins.rs:244]: yep2, did tried it, care to check it out?

**Comment by geofmureithi** [comment]
About the classes macro, I am curious if any crates work out of the box? IMO we should be able to reuse crates that exist. I will check out the current implementations tonight and get back.

**Pull request closed without merging**